// gatsby-source-contentful/fragments.js
"use strict";

exports.__esModule = true;
/* eslint-disable */
var contentfulAssetResolutions = exports.contentfulAssetResolutions = graphql`
  fragment GatsbyContentfulResolutions on ContentfulResolutions {
    base64
    width
    height
    src
    srcSet
  }
`;

var contentfulAssetResolutionsTracedSVG = exports.contentfulAssetResolutionsTracedSVG = graphql`
  fragment GatsbyContentfulResolutions_tracedSVG on ContentfulResolutions {
    tracedSVG
    width
    height
    src
    srcSet
  }
`;

var contentfulAssetResolutionsNoBase64 = exports.contentfulAssetResolutionsNoBase64 = graphql`
  fragment GatsbyContentfulResolutions_noBase64 on ContentfulResolutions {
    width
    height
    src
    srcSet
  }
`;

var contentfulAssetResolutionsPreferWebp = exports.contentfulAssetResolutionsPreferWebp = graphql`
  fragment GatsbyContentfulResolutions_withWebp on ContentfulResolutions {
    base64
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;

var contentfulAssetResolutionsPreferWebpNoBase64 = exports.contentfulAssetResolutionsPreferWebpNoBase64 = graphql`
  fragment GatsbyContentfulResolutions_withWebp_noBase64 on ContentfulResolutions {
    width
    height
    src
    srcSet
    srcWebp
    srcSetWebp
  }
`;

var contentfulAssetSizes = exports.contentfulAssetSizes = graphql`
  fragment GatsbyContentfulSizes on ContentfulSizes {
    base64
    aspectRatio
    src
    srcSet
    sizes
  }
`;

var contentfulAssetSizesTracedSVG = exports.contentfulAssetSizesTracedSVG = graphql`
  fragment GatsbyContentfulSizes_tracedSVG on ContentfulSizes {
    tracedSVG
    aspectRatio
    src
    srcSet
    sizes
  }
`;

var contentfulAssetSizesNoBase64 = exports.contentfulAssetSizesNoBase64 = graphql`
  fragment GatsbyContentfulSizes_noBase64 on ContentfulSizes {
    aspectRatio
    src
    srcSet
    sizes
  }
`;

var contentfulAssetSizesPreferWebp = exports.contentfulAssetSizesPreferWebp = graphql`
  fragment GatsbyContentfulSizes_withWebp on ContentfulSizes {
    base64
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;

var contentfulAssetSizesPreferWebpNoBase64 = exports.contentfulAssetSizesPreferWebpNoBase64 = graphql`
  fragment GatsbyContentfulSizes_withWebp_noBase64 on ContentfulSizes {
    aspectRatio
    src
    srcSet
    srcWebp
    srcSetWebp
    sizes
  }
`;
